Status check for test and utility methods read_file and write_file

shahzebsiddiqui opened the pull request:
This PR successfully implements the status key implementation in the framework related to schema change addressed in https://github.com/buildtesters/schemas/pull/24 

The commit message along with comments should explain the logic on how i implemented the test_state (PASS/FAIL) which is used to count the number of pass and failed test.

I have ran three test, one with status check using returncode, one with regex and one using both. 

The three tests were committed to tutorials repo see commit https://github.com/buildtesters/tutorials/commit/5406b84e048557fb2d81b4de0febe847ca862164 

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b system/systemd.yml -b system/selinux.yml  -b apps/openacc/vecadd.yml

            Discovered Buildspecs

/mxg-hpc/users/ssi29/tutorials/apps/openacc/vecadd.yml
/mxg-hpc/users/ssi29/tutorials/system/systemd.yml
/mxg-hpc/users/ssi29/tutorials/system/selinux.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path   
________________________________________________________________________________________________________________________
vecadd                         vecadd_gnu                     PASS                           /mxg-hpc/users/ssi29/tutorials/apps/openacc/vecadd.yml
systemd                        systemd_default_target         PASS                           /mxg-hpc/users/ssi29/tutorials/system/systemd.yml
selinux                        selinux_disable                PASS                           /mxg-hpc/users/ssi29/tutorials/system/selinux.yml


============================================================
                        Test summary
============================================================
Executed 3 tests
Passed Tests: 3/3 Percentage: 100.000%
Failed Tests: 0/3 Percentage: 0.000%
```

Here is a snippet from the logs to give idea whats going on

**selinux.yml** - This one performs both regex and return code check

```
2020-04-22 11:44:32,122 [base.py:535 - run_tests() ] - [DEBUG] Return code: 0 for test: /mxg-hpc/users/ssi29/tutorials/.buildtest/selinux/selinux_disable.sh
2020-04-22 11:44:32,122 [base.py:555 - run_tests() ] - [DEBUG] Conducting Return Code check
2020-04-22 11:44:32,122 [base.py:558 - run_tests() ] - [DEBUG] Status Return Code: 0   Result Return Code: 0
2020-04-22 11:44:32,122 [base.py:567 - run_tests() ] - [DEBUG] Conducting Regular Expression check
2020-04-22 11:44:32,122 [base.py:471 - check_regex() ] - [DEBUG] Detected regex stream 'stdout' so reading output file: /mxg-hpc/users/ssi29/tutorials/.buildtest/selinux/run/selinux_disable_04-22-2020-11-44.out
2020-04-22 11:44:32,123 [base.py:482 - check_regex() ] - [DEBUG] Applying re.search with exp: ^SELINUX=disabled$
2020-04-22 11:44:32,123 [base.py:572 - run_tests() ] - [INFO] ReturnCode Match: True Regex Match: True

```

**systemd.yml** - this performs return code check only

```
2020-04-22 11:44:32,077 [base.py:513 - run_tests() ] - [DEBUG] Running Test via command: /bin/bash /mxg-hpc/users/ssi29/tutorials/.buildtest/systemd/systemd_default_target.sh
2020-04-22 11:44:32,094 [base.py:530 - run_tests() ] - [DEBUG] Writing run output to file: /mxg-hpc/users/ssi29/tutorials/.buildtest/systemd/run/systemd_default_target_04-22-2020-11-44.out
2020-04-22 11:44:32,095 [base.py:534 - run_tests() ] - [DEBUG] Writing run error to file: /mxg-hpc/users/ssi29/tutorials/.buildtest/systemd/run/systemd_default_target_04-22-2020-11-44.err
2020-04-22 11:44:32,095 [base.py:535 - run_tests() ] - [DEBUG] Return code: 0 for test: /mxg-hpc/users/ssi29/tutorials/.buildtest/systemd/systemd_default_target.sh
2020-04-22 11:44:32,095 [base.py:555 - run_tests() ] - [DEBUG] Conducting Return Code check
2020-04-22 11:44:32,095 [base.py:558 - run_tests() ] - [DEBUG] Status Return Code: 0   Result Return Code: 0
2020-04-22 11:44:32,095 [base.py:572 - run_tests() ] - [INFO] ReturnCode Match: True Regex Match: True
2020-04-22 11:44:32,095 [base.py:70 - __init__() ] - [DEBUG] buildtest found the available schema: 
```

**vecadd.yml** - this performs regex check only

```
2020-04-22 11:44:31,692 [base.py:513 - run_tests() ] - [DEBUG] Running Test via command: /bin/bash /mxg-hpc/users/ssi29/tutorials/.buildtest/vecadd/vecadd_gnu.sh
2020-04-22 11:44:32,068 [base.py:530 - run_tests() ] - [DEBUG] Writing run output to file: /mxg-hpc/users/ssi29/tutorials/.buildtest/vecadd/run/vecadd_gnu_04-22-2020-11-44.out
2020-04-22 11:44:32,069 [base.py:534 - run_tests() ] - [DEBUG] Writing run error to file: /mxg-hpc/users/ssi29/tutorials/.buildtest/vecadd/run/vecadd_gnu_04-22-2020-11-44.err
2020-04-22 11:44:32,069 [base.py:535 - run_tests() ] - [DEBUG] Return code: 0 for test: /mxg-hpc/users/ssi29/tutorials/.buildtest/vecadd/vecadd_gnu.sh
2020-04-22 11:44:32,069 [base.py:567 - run_tests() ] - [DEBUG] Conducting Regular Expression check
2020-04-22 11:44:32,069 [base.py:471 - check_regex() ] - [DEBUG] Detected regex stream 'stdout' so reading output file: /mxg-hpc/users/ssi29/tutorials/.buildtest/vecadd/run/vecadd_gnu_04-22-2020-11-44.out
2020-04-22 11:44:32,069 [base.py:482 - check_regex() ] - [DEBUG] Applying re.search with exp: ^final result: 1.000000$
2020-04-22 11:44:32,069 [base.py:572 - run_tests() ] - [INFO] ReturnCode Match: True Regex Match: True

```

Finally, if status is not defined buildtest works as normal, here is an example run without any status check. We are running the test https://github.com/buildtesters/tutorials/blob/master/apps/serial/hello.yml

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b apps/serial/hello.yml

            Discovered Buildspecs

/mxg-hpc/users/ssi29/tutorials/apps/serial/hello.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path   
________________________________________________________________________________________________________________________
hello                          fortan_hello                   PASS                           /mxg-hpc/users/ssi29/tutorials/apps/serial/hello.yml
hello                          c_hello                        PASS                           /mxg-hpc/users/ssi29/tutorials/apps/serial/hello.yml
hello                          cplusplus_hello                PASS                           /mxg-hpc/users/ssi29/tutorials/apps/serial/hello.yml


============================================================
                        Test summary
============================================================
Executed 3 tests
Passed Tests: 3/3 Percentage: 100.000%
Failed Tests: 0/3 Percentage: 0.000%

```

I have tested locally by tweaking returncode and regex to see if PASS/FAIL is working and it was successful. Feel free to test this by cloning the ``tutorials`` repo and set ``WORK`` environment to parent directory where you cloned tutorials and it should work.

shahzebsiddiqui committed:
add implementation to report test state based on 'status' object defined
in Buildspec.
Add method check_returncode and check_regex to BuilderBase class that
return boolean True/False.
If status is not defined, we perform check based on returncode from
test. If its a 0, test_state=PASS otherwise set to FAIL.
We add the following keys to self.metadata: 'outfile', 'errfile',
'TEST_STATE'.
In buildtest/menu/build.py we count PASS/FAIL test based on
self.metadata['TEST_STATE']

shahzebsiddiqui committed:
make use of self.metadata['outfile'] and self.metadata['errfile'] when
writing output and error stream to files.

shahzebsiddiqui committed:
add docstring for check_returncode and check_regex method in BuilderBase

shahzebsiddiqui committed:
fix black issue

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -431,6 +432,60 @@ def run(self):
+            self.logger.debug(
+                f"Detected regex stream 'stdout' so reading output file: {self.metadata['outfile']}"
+            )
+            with open(self.metadata["outfile"], "r") as fd:
```
Would it be worth to have a utils. read_file and write_file to be used here? They are relatively simple functions, but would be one line instead of two :)

shahzebsiddiqui replied:
If i understand you want two method in buildtest/utils/file.py ``read_file`` and ``write_file``. that makes sense, 1 one line is better than 2. I can get this implemented in this PR.

vsoch replied:
Yes! You've used it twice in this PR, so it makes sense to add. You can probably grep for other instances of open( to find other occurrences too.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -431,6 +432,60 @@ def run(self):
+            :rtype: bool
+        """
+
+        if regex["stream"] == "stdout":
```
Can you be sure "stream" is a field of regex? What if it's 1. not there, or 2. there but not matching?

shahzebsiddiqui replied:
I made sure ``stream`` and ``exp`` are required https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json#L73 
Also enum set for ``stream`` so its not possible https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json#L69
And this is also tested in the following

https://github.com/buildtesters/schemas/blob/master/.github/tests/invalid/script/0.0.1/invalid_script_v0.0.1.yml#L38
https://github.com/buildtesters/schemas/blob/master/.github/tests/invalid/script/0.0.1/invalid_script_v0.0.1.yml#L47

That should clarify any doubts :)

vsoch replied:
Awesome! Yes, clarified.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -431,6 +432,60 @@ def run(self):
+
+        self.logger.debug(f"Applying re.search with exp: {regex['exp']}")
+        # perform a regex search based on value of 'exp' key defined in Buildspec with content file (output or error)
+        if re.search(regex["exp"], content):
```
This one too - how are we sure this is provided?

shahzebsiddiqui replied:
I think this is resolved with my response in https://github.com/buildtesters/buildtest/pull/286#discussion_r413178872 

There is a test for this too https://github.com/buildtesters/schemas/blob/master/.github/tests/invalid/script/0.0.1/invalid_script_v0.0.1.yml#L47

I believe the ``validate`` method will complain if this key is not present. Right now i checked if ``stream`` is passed without ``exp`` i didn't check the other way around but i take it that would be redundant test for ``required``.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +521,77 @@ def run_tests(self, testfile):
+
         # write output of test to .out file
-        with open(run_output_file + ".out", "w") as fd:
+        with open(self.metadata["outfile"], "w") as fd:
```
We could use the utils functions here too :) And it would be good to add tests.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +521,77 @@ def run_tests(self, testfile):
+            regex_match = True
+
+            # check returncode from result matches value defined in Buildspec recipe, self.check_returncode returns a bool
+            if "returncode" in status.keys():
```
You don't need "keys" here, you can just say if key in dict.

shahzebsiddiqui replied:
Good call, i fixed this now in https://github.com/shahzebsiddiqui/buildtest-framework/commit/1ba7ae328775f96715281dfb79887b4b195e6dad.

vsoch reviewed (changes requested):
Looking good! I addressed some issues with the function, and specifically let's:
 - add util functions for writing / reading files with tests
 - can we add tests for these return statuses / what you showed?

Just to verify in some of the output you showed - there was an __init__ function showed in the output for the second that wasn't shown for the first and third. Was this just an extra copy paste line or some actual difference / error? (related to the schema found but seemed like an incomplete message).

shahzebsiddiqui committed:
adding utility method read_file and write_file for reading and writing
files. Refactor Builderbase to use read_file and write_file when
writing output and error file.
Adding regression test for checking the two utility methods.

shahzebsiddiqui committed:
forgot to address an issue raised by @vsoch regarding key search
in status dictionary.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,80 @@ def resolve_path(path):
+    """
+
+    # type check filepath to ensure its a string, if not return None
+    if not isinstance(filepath,str):
```
I think this will need black formatting, should be a space after filepath.

shahzebsiddiqui replied:
that's strange ill fix this, i have pre-commit hook enabled but i run into these issues. Don't worry if you see a failure on black. Ill fix those in subsequent commits if it happens.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,80 @@ def resolve_path(path):
+        return None
+
+    content = []
+    fd = open(filepath, "r")
```
This is overly complicated - we just need:
```
with open(filepath, 'r') as fd:
    content = fd.read()
```
This parsing of lines doesn't make sense. If we wanted to read lines we could do readline() but it's better to return the raw content of the file and let the parsing function decide what to do. For example, if we used this and piped into a function to parse yaml, it would be all kinds of weird to get a list of strings.

shahzebsiddiqui replied:
ahhh, i was kinda following the principles of read and write by line similar to ``BuildTestCommand`` . If you wanted a simple as that, i could have saved some time :(
Would you like the return type for ``read_file`` be a string or a list.

vsoch replied:
Sorry for the misunderstanding! I literally meant changing those 2 lines into a function since we re-use, and adding a check (for read) that the file exists first, and (for write) that the directory to write exists first.

Generally, we don't expect special / custom parsing, so you'd want to use read() and just return raw content, same idea for write.

shahzebsiddiqui replied:
ok if i understand i will create two methods ``read`` and ``write`` with the suggestion you proposed.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,80 @@ def resolve_path(path):
+        return None
+
+    # if content was passed as a string, let's convert to list
+    if isinstance(content, str):
```
This also doesn't make sense - just take the content given from the user, and check that it's a string, and write it to the file. If the user wants newlines, etc., this is up for him/her to pass.

shahzebsiddiqui replied:
just to clarify we should only accept ``str`` type to ``write_file`` when passing the content of file. If it's not a string we can return ``None``

vsoch replied:
I think it's fine to assume that a string is passed, and let the error be raised if something else is. I would just check that the filename exists for read, and that the dirname exists for write.

shahzebsiddiqui replied:
So i initially allowed an input of list because ``BuildTestCommand.execute`` returns output, error as a ``list``. These are then invoked to ``write_file``.
Just to confirm you would want input content to be in ``string``, if it's list would you want ``write_test`` to do nothing or convert to string and try to write file. Btw, i initially assumed input could be string or list so i convert it.

vsoch replied:
`write_file("\n".join(output))`

shahzebsiddiqui replied:
Unrelated to this, but i was testing locally something is wrong with ``BuildTestCommand`` it was working before now it's only getting first line of file and i can reproduce this on the client. The input content ``out`` and ``err`` are invoked via ``write_file`` so i was trying to figure out if this was an issue with my function, seems like it's coming from ``BuildTestCommand``

```
>>> from buildtest.utils.command import BuildTestCommand
>>> cmd = BuildTestCommand("python .buildtest/python-circle/circle_area.py")
>>> out,err = cmd.execute()
>>> out
['Circle Radius  2\n']

>>> os.system("python .buildtest/python-circle/circle_area.py")
Circle Radius  2
Area of circle  12.566370614359172
Block
0

```

shahzebsiddiqui replied:
This issue is present in ``devel`` branch, ill open an issue basically its reading first line and stopping.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,80 @@ def resolve_path(path):
+    # None.
+    filepath = os.path.expanduser(filepath)
+    filepath = os.path.expandvars(filepath)
+    filepath = os.path.realpath(filepath)
```
This is also not scoped to the function - we simply need to check if it's a string, and if the directory name to write it to exists, and then write it :) These functions should be 3-6 lines at most!

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,80 @@ def resolve_path(path):
+        return None
+    
+    fd = open(filepath, "w")
+    # process each line and write to file
```
Don't forget that it's best practice to use with:

shahzebsiddiqui committed:
add method read, write which are doing the read and write operation.
Refactor read_file and write_file.
We expect input content for writing to be a string.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,76 @@ def resolve_path(path):
         return real_path
+
+
+def read(filepath):
```
Please call this "read_file" and then "write_file."

shahzebsiddiqui replied:
this is fixed now in https://github.com/shahzebsiddiqui/buildtest-framework/commit/1fbe34ca33da9f65698e68e153a7315491a02979

shahzebsiddiqui committed:
remove method 'read' and  'write' and merge them to 'read_file' and
'write_file'

shahzebsiddiqui committed:
applying black

Signed-off-by: Shahzeb Siddiqui <shahzebmsiddiqui@gmail.com>

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -431,6 +432,61 @@ def run(self):
+            :rtype: bool
+        """
+
+        if recipe_returncode == result_returncode:
```
You can just return `recipe_returncode == result_returncode`

shahzebsiddiqui replied:
got it, i fixed this locally.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +522,81 @@ def run_tests(self, testfile):
+                    status["returncode"], result["RETURN_CODE"]
+                )
+
+            # check regex expression in Buildspec with output or error stream. self_check_regex returns a boolean (True/False)
```
this line / documentation should go with the check_regex function.

shahzebsiddiqui replied:
fixed locally.

shahzebsiddiqui committed:
update comments near method invocation 'self.check_regex',
'self.check_returncode'
Remove if conditions in self.check_returncode

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -431,6 +432,58 @@ def run(self):
+        content = "\n".join(content)
+        self.logger.debug(f"Applying re.search with exp: {regex['exp']}")
+        # perform a regex search based on value of 'exp' key defined in Buildspec with content file (output or error)
+        if re.search(regex["exp"], content):
```
This could also be shortened to return `re.search(..) != None` if you want to use fewer lines :)

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +519,82 @@ def run_tests(self, testfile):
-                    self.config_name, self.name, "PASSED", self.buildspec
+        status = self.recipe.get("status")
+
+        test_state = ""
```
Any reason to not have it start as FAIL up here, and then just update to PASS given passing conditions? Right now there is an if -> else flow, and the else always sets it to FAIL - why not just start like that? A default state is to FAIL until there is evidence of passing. That will also simplify some of the logic below.

shahzebsiddiqui replied:
good call, I see your point, set **test_state="FAIL"** and remove the else conditions. Ill fix this.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +519,82 @@ def run_tests(self, testfile):
+            if returncode_match and regex_match:
+                test_state = "PASS"
+            else:
+                test_state = "FAIL"
```
If this is default (FAIL), you don't need the else: fail. It will already be set and you just need to set to be success given the first condition. It's the same for line ~587.

shahzebsiddiqui replied:
okay i believe this is fixed in https://github.com/shahzebsiddiqui/buildtest-framework/commit/b4750fdab1dd4367d448f7c51dd21e3098c3e92d

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,66 @@ def resolve_path(path):
+
+    # type check filepath to ensure its a string, if not return None
+    if not isinstance(filepath, str):
+        return None
```
You have the return type here a list? I would also add a note that returning None indicates some problem with the file. Is there any reason not to exit on error given the return is None? We want to make sure that any Nones returned are properly handled.

shahzebsiddiqui replied:
this is checking is to ensure ``filepath``   is of type string. I could with System Error on this if that makes sense.

vsoch replied:
That makes sense, yes.

shahzebsiddiqui replied:
this is resolved but with a SystemError message see https://github.com/shahzebsiddiqui/buildtest-framework/commit/e4ad6f2188b00e085e92d94b225b8587e1c03cdf

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,66 @@ def resolve_path(path):
+        :rtype: list
+    """
+
+    # type check filepath to ensure its a string, if not return None
```
You can combine these first two conditions into one if --> return statement.

shahzebsiddiqui replied:
I am not following, you want me to change the comment or are you referring to the two conditions

```
    # type check filepath to ensure its a string, if not return None
    if not isinstance(filepath, str):
        return None

    # ensure content is of type string
    if not isinstance(content, str):
        return None
```

vsoch replied:
```python
if not isinstance(filepath, str) and not isinstance(content, str):
    return None
```

shahzebsiddiqui replied:
I changed the behavior with meaningful exception messages, i kept them separate so i can get coverage for each if condition with the appropriate test.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +143,66 @@ def resolve_path(path):
+    if not isinstance(content, str):
+        return None
+
+    # if filepath is an actual file, let's not write to file and return None.
```
This isn't expected behavior - I would want to give a write file function a file (that already might exist) and have it over-written. If you are concerned about this, then there could be a check in the calling function, but it shouldn't be here with the util (and then silently not update the file that exists, that sounds like a bad silent bug to me!)

shahzebsiddiqui replied:
Okay i removed this check so you can overwrite to file.  This is added in https://github.com/shahzebsiddiqui/buildtest-framework/commit/e4ad6f2188b00e085e92d94b225b8587e1c03cdf#diff-b94bc4259405fb65c0cd50bf50dc3bc1L200

shahzebsiddiqui committed:
set test_state=FAIL as default value.
self.check_regex was spliting each string by character which was not
right.

shahzebsiddiqui committed:
applying black

shahzebsiddiqui committed:
refactor read_file and write_file so that we raise error when
input type is invalid. Rewrite regtest for read_file, write_file based
on recent change.
Update docstrings

shahzebsiddiqui committed:
Add print statement, and increase coverage for file
buildtest/utils/file.py
Update header describing file utils/file.py

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -431,6 +432,57 @@ def run(self):
+            )
+            content = read_file(self.metadata["errfile"])
+
+        # convert list to string
```
This should be removed, generally it’s not good practice to leave commented code.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +518,78 @@ def run_tests(self, testfile):
+        err = "\n".join(err)
+
+        write_file(self.metadata["outfile"], out)
+        self.logger.debug(f"Writing run output to file: {self.metadata['outfile']}")
```
You’d want this line before attempting the write in case something goes wrong it will appear in the log.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +518,78 @@ def run_tests(self, testfile):
-            fd.write("\n".join(err))
-        self.logger.debug(f"Writing run error to file: {run_output_file + '.err'}")
+        write_file(self.metadata["errfile"], err)
+        self.logger.debug(f"Writing run error to file: {self.metadata['errfile']}")
```
Same here.

vsoch commented on buildtest/buildsystem/base.py:
```diff
@@ -466,31 +518,78 @@ def run_tests(self, testfile):
+                    % (status["returncode"], result["RETURN_CODE"])
                 )
+                # self.check_returncode checks if test returncode matches returncode specified in Buildspec.
+                # The return value is a bool
```
Why is there an entire function to check for equality of two values (one line)?

shahzebsiddiqui replied:
I foresee we can invoke or override method in future for example slurm job codes would need to be retrieved first before comparing with value in Buildspec. Some of these methods in BuilderBase could be better suited to be defined in ExecutorBase  class.

vsoch replied:
I think this method would make sense only in the case that a return code was retrieved internally. Since we have to provide both values, there is no benefit /use to doing this as the user will always need to provide those two values (and just as easily could compare them.) For your use case that you mention, I would suggest having the check for success happen on the level of a specific kind of executor (e.g., slurm). If the function is to be meaningful here, it should use something internally. If this is truly needed / to be overridden by a custom executor, that should be added when the executor is written, and not before hand (now.)

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +138,62 @@ def resolve_path(path):
+        :rtype: str
+    """
+
+    # type check filepath to ensure its a string, if not return None
```
Docstring here needs update.

vsoch commented on buildtest/utils/file.py:
```diff
@@ -143,3 +138,62 @@ def resolve_path(path):
+        :type content: str, required
+    """
+
+    # type check filepath to ensure its a string, if not return None
```
Docstring needs update.

shahzebsiddiqui committed:
addressing comments by @vsoch

shahzebsiddiqui committed:
update docstring for read_file and write_file with complete description
outlining behavior of method along with description of Exceptions raised

vsoch reviewed (approved):
If you are ok with decrease in coverage, I have differences in opinion on some of the changes, but it passes the bar for basic review, so feel free to merge.
